ctru-rs/src/applets/error.rs

```diff
@@ -0,0 +1,149 @@
+
+        Self { state }
+    }
+
+    /// Sets the error text to display.
+    #[doc(alias = "errorText")]
+    pub fn set_text(&mut self, text: &str) {
+        for (idx, code_unit) in text
+            .encode_utf16()
+            .chain(std::iter::once(0))
+            .take(self.state.Text.len() - 1)
+            .enumerate()
+        {
+            self.state.Text[idx] = code_unit;
+        }
+    }
+
+    /// Launches the error applet.
+    #[doc(alias = "errorDisp")]
+    pub fn launch(&mut self, _apt: &Apt, _gfx: &Gfx) -> Result<(), Error> {
+        unsafe { ctru_sys::errorDisp(self.state.as_mut()) }
+
+        match self.state.returnCode {
+            ctru_sys::ERROR_NONE | ctru_sys::ERROR_SUCCESS => Ok(()),
+            ctru_sys::ERROR_NOT_SUPPORTED => Err(Error::NotSupported),
+            ctru_sys::ERROR_HOME_BUTTON => Err(Error::HomePressed),
+            ctru_sys::ERROR_POWER_BUTTON => Err(Error::PowerPressed),
+            ctru_sys::ERROR_SOFTWARE_RESET => Err(Error::ResetPressed),
+            _ => Err(Error::Unknown),
+        }
+    }
+
+    /// Launches the error applet without requiring an [`Apt`] or [`Gfx`] handle.
+    ///
+    /// # Safety
+    ///
+    /// Causes undefined behavior if the aforementioned services are not actually active when the applet launches.
+    unsafe fn launch_unchecked(&mut self) {
+        unsafe { ctru_sys::errorDisp(self.state.as_mut()) };
+    }
+}
+
+/// Sets a custom panic hook that uses the error applet to display panic messages. You can also choose to have
+/// messages printed over stderr along with the pop-up display.
+///
+/// If the `Gfx` service is not initialized during a panic, the error applet will not be displayed and the panic
+/// message will be printed over stderr.
+pub fn set_panic_hook(use_stderr: bool) {
+    use crate::services::gfx::GFX_ACTIVE;
+
+    use std::io::Write;
+    use std::sync::TryLockError;
+
+    std::panic::set_hook(Box::new(move |panic_info| {
+        let _apt = Apt::new();
+
+        let thread = std::thread::current();
+
+        let name = thread.name().unwrap_or("<unnamed>");
+
+        // If we get a `WouldBlock` error, we know that the `Gfx` service has been initialized.
+        // Otherwise fallback to printing over stderr.
+        if let Err(TryLockError::WouldBlock) = GFX_ACTIVE.try_lock() {
+            if use_stderr {
+                print_to_stderr(name, panic_info);
+            }
+
+            let payload = format!("thread '{name}' {panic_info}");
+
+            let mut popup = PopUp::new(Kind::Top);
+
+            popup.set_text(&payload);
+
+            unsafe {
+                popup.launch_unchecked();
+            }
+        } else {
+            print_to_stderr(name, panic_info);
+        }
+    }));
+
+    fn print_to_stderr(name: &str, panic_info: &std::panic::PanicInfo) {
+        let mut stderr = std::io::stderr().lock();
+
+        let _ = writeln!(stderr, "thread '{name}' {panic_info}");
+    }
```

Comment:
Instead of `use_stderr` could we provide an option to use the default hook or not? something like
```rs
pub fn set_panic_hook(call_existing_hook: bool) {
	let old_hook = std::panic::take_hook();

	std::panic::set_hook(Box::new(move |panic_info| {
		if let Err(TryLockError::WouldBlock) = GFX_ACTIVE.try_lock() {
			// ...
		}

		if call_existing {
			old_hook(panic_info);
		}
	})
}
```